tests/test_llama.py

```diff
@@ -0,0 +1,176 @@
+    os.setpgrp()  # create new process group, become its leader
+    atexit.register(exit_with_children)  # kill all children processes when this process exits
+
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+
+    # Read and print output in real-time
+    while True:
+        line = process.stdout.readline()
+        if process.poll() is not None and line == b"":
+            break
+        if line:
+            print(line.decode("utf-8"), end="")
+            # for all iterations >= CHECK_ITERATION, loss should be below EXPECTED_LOSS
+            if re.search(r"iteration: (\d+) / ", line.decode("utf-8")):
+                if int(re.search(r"iteration: (\d+) / ", line.decode("utf-8")).group(1)) >= CHECK_ITERATION:
+                    loss = extract_loss(line)
+                    assert loss < EXPECTED_LOSS
+            # for iteration = CHECK_ITERATION_END, loss should be below EXPECTED_LOSS_END
+            if re.search(rf"iteration: {CHECK_ITERATION_END} / ", line.decode("utf-8")):
+                loss = extract_loss(line)
+                assert loss < EXPECTED_LOSS_END
+
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+
+# also run the tiny llama example. Only want to assert it can be ran.
+def test_tiny_llama():
+    # create config file
+    cmd = f"python {TINY_LLLAMA_CREATE_CONFIG_FILE}"
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    # Read and print output in real-time
+    while True:
+        line = process.stdout.readline()
+        if process.poll() is not None and line == b"":
+            break
+        if line:
+            print(line.decode("utf-8"), end="")
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+    # run training
+    # set DISABLE_FLASH_ATTENTION=1 to replace flash attention implementations
+    cmd = f'DISABLE_FLASH_ATTENTION=1 FI_PROVIDER="efa" CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun --nproc_per_node={NUM_GPUS} --rdzv_endpoint=localhost:29800  {TRAIN_SCRIPT} --config-file {TINY_LLLAMA_CONFIG_FILE}'
+    os.setpgrp()  # create new process group, become its leader
+    atexit.register(exit_with_children)  # kill all children processes when this process exits
+
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    while True:
+        line = process.stdout.readline()
+        if process.poll() is not None and line == b"":
+            break
+        if line:
+            print(line.decode("utf-8"), end="")
+
+    process.wait()  # Wait for the process to finish
+    assert process.returncode == 0
+
+
+if __name__ == "__main__":
+    # create config file
+    cmd = f"python {CREATE_CONFIG_FILE}"
+    process = subprocess.Popen(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
```

Comment:
Can we run it through a function like test_train_tiny_llama, so that vscode pytest can run all the tests including this one (pytest ./tests)? So we don't need to run it separately through the command line

Reply:
If my understanding is correct, we can already do it without adding a function. pytest just automatically run all the functions starting with "test".